### ISSUE TITLE
Question regarding debugging Authenticated vs. Unauthenticated GDAX requests

### ISSUE THREAD

**Issue by kerryknight** [coinbase/gdax-tt#34]
**TLDR: Can anyone give me a slight bit of direction where it seems my issue is stemming from (the gdax-tt source, my local system setup or the gdax WS API) as I get failed trades when following the tutorial guide and without entering gdax authentication keys but then I get websocket connection closed errors when I do provide my keys manually?**

Hi, I'm trying to get my feet wet by testing out the `t005_alertTrader.ts` tutorial.  I'm trying to figure out if I'm running into a gdax-tt source, a local system setup or a gdax WS API issue, as I'm unable to complete a basic test trade with valid api keys.  I have tried 2 different api keys I created on gdax, both with full permissions.  I have added ENVs via `dotenv` but I'm unsuccessful completing a basic test trade whether I add my ENVs to the tutorial script or not.  I am able to at least connect and get responses from the GDAX WS API when I leave the tutorial script untouched as mentioned in the [tutorial's advice](https://coinbase.github.io/gdax-tt/gtt_tutorials_triggers.html) of:

`"You can set auth: null to just use the defaults, which since you have your GDAX API keys set in the environment, will automatically use those and receive authenticated messages (nice if you want to confim when your trades are filled)."`

However, whenever the script sees a market opportunity and attempts to make a trade, it fails, presumably because my ENVs are not actually getting passed like the tutorial says.  I get output similar to this:
```
MacBook-Pro:gdax-tt goheelsgo$ ts-node tutorials/t005_alertTrader.ts
2017-09-26T13:25:53.072Z - info: Creating new Websocket connection to wss://ws-feed.gdax.com 
2017-09-26T13:25:53.090Z - debug: Reconnecting to wss://ws-feed.gdax.com  in 0 seconds...
2017-09-26T13:25:53.090Z - debug: Closing existing socket prior to reconnecting to wss://ws-feed.gdax.com
2017-09-26T13:25:53.331Z - debug: Connection to wss://ws-feed.gdax.com  has been established.
2017-09-26T13:25:53.331Z - debug: Sending {"type":"subscribe","product_ids":["BTC-USD"],"channels":["ticker","heartbeat"]} message to WS server
Price:   3954.500 | Bid:   3954.490 | Ask:   3954.500 | sequence: undefined
Price:   3954.500 | Bid:   3954.490 | Ask:   3954.500 | sequence: 21336971
.
.
.
2017-09-26T13:26:17.957Z - error: Placing order failed type=order, time=null, productId=BTC-USD, orderType=market, side=buy, size=0.005, reason=A GDAX API request failed. Bad Request
(node:6783) UnhandledPromiseRejectionWarning: Unhandled promise rejection (rejection id: 3): Error: A GDAX API request failed. Bad Request
(node:6783) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.
Price:   3954.010 | Bid:   3954.000 | Ask:   3954.010 | sequence: 21336983
```
Now, when i set my ENVs in the tutorial script:
```
const options: GDAXFeedConfig = {
    logger: logger,
    auth: { 
        key: process.env.GDAX_KEY, 
        secret: process.env.GDAX_SECRET, 
        passphrase: process.env.GDAX_PASSPHRASE 
    }, // use public feed
    channels: ['ticker'],
    wsUrl: GDAX_WS_FEED,
    apiUrl: GDAX_API_URL
};
```
It appears to pass them in the config options, yet I get a websocket connection closed by external party 1006 error over and over again, presumably due to a timeout?  

```
MacBook-Pro:gdax-tt goheelsgo$ ts-node tutorials/t005_alertTrader.ts
2017-09-26T13:17:53.197Z - info: Creating new Websocket connection to wss://ws-feed.gdax.com (authenticated)
2017-09-26T13:17:53.217Z - debug: Reconnecting to wss://ws-feed.gdax.com (authenticated) in 0 seconds...
2017-09-26T13:17:53.217Z - debug: Closing existing socket prior to reconnecting to wss://ws-feed.gdax.com
2017-09-26T13:17:53.446Z - debug: Connection to wss://ws-feed.gdax.com (authenticated) has been established.
2017-09-26T13:17:53.447Z - debug: Sending {"type":"subscribe","product_ids":["BTC-USD"],"channels":["ticker","heartbeat"],"signature":"<REDACTED>","key":"<REDACTED>","timestamp":"1506431873.447","passphrase":"<REDACTED>"} message to WS server
2017-09-26T13:18:53.219Z - info: The websocket feed to wss://ws-feed.gdax.com (authenticated) has been closed by an external party. We will reconnect in 5 seconds code=1006, reason=
2017-09-26T13:18:53.220Z - debug: Reconnecting to wss://ws-feed.gdax.com (authenticated) in 5 seconds...
2017-09-26T13:18:58.364Z - debug: Connection to wss://ws-feed.gdax.com (authenticated) has been established.
2017-09-26T13:18:58.365Z - debug: Resubscribing to BTC-USD...
2017-09-26T13:18:58.365Z - debug: Sending {"type":"subscribe","product_ids":["BTC-USD"],"channels":["ticker","heartbeat"],"signature":"<REDACTED>","key":"<REDACTED>","timestamp":"1506431938.365","passphrase":"<REDACTED>"} message to WS server
2017-09-26T13:18:58.366Z - debug: Sending {"type":"subscribe","product_ids":["BTC-USD"],"channels":["ticker","heartbeat"],"signature":"<REDACTED>","key":"<REDACTED>","timestamp":"1506431938.366","passphrase":"<REDACTED>"} message to WS server
2017-09-26T13:18:58.367Z - debug: Reconnection to BTC-USD successful
2017-09-26T13:19:57.911Z - info: The websocket feed to wss://ws-feed.gdax.com (authenticated) has been closed by an external party. We will reconnect in 5 seconds code=1006, reason=
```

The trade I'm attempting to make is around $15-20 and I've got $50 in my GDAX in USD, so I'd think I wouldn't have any issue with minimums.  I'm a bit at a loss currently as Node is not my forte--can anyone give me a bit of direction where it seems my issue is stemming from?  
```
MacBook-Pro:gdax-tt goheelsgo$ ts-node -v
ts-node v3.3.0
node v8.4.0
typescript v2.5.2
```

**Comment by CjS77** [comment]
I've checked tutorial 5 again, and the trade is placed.

BTW, Are you using v0.1.16? There were a few auth issues that have been ironed out in the latest version.

**Comment by CjS77** [comment]
Regarding the disconnects, GDAX did have some instability a few days ago and you might have been testing during that period. If you run with the latest GTT version now, do you still have the same issues?

**Comment by kerryknight** [comment]
Awesome, thanks for double-checking.  Looks like I'm indeed on v0.1.16.
```
├─┬ gdax-trading-toolkit@0.1.16
│ ├── @types/bignumber.js@4.0.2 deduped
│ ├── @types/bintrees@1.0.2 deduped
│ ├── @types/commander@2.3.31 deduped
│ ├── @types/crypto-js@3.1.32 deduped
│ ├── @types/mocha@2.2.33 deduped
│ ├── @types/nock@8.2.0 deduped
│ ├── @types/node@6.0.51 deduped
│ ├── @types/query-string@3.0.30 deduped
│ ├── @types/simple-mock@0.0.27 deduped
│ ├── @types/superagent@2.0.35 deduped
│ ├── @types/uuid@2.0.29 deduped
│ ├── @types/winston@2.2.0 deduped
│ ├── @types/ws@0.0.37 deduped
│ ├── bignumber.js@4.0.2 deduped
│ ├── bintrees@1.0.1 deduped
│ ├── ccxt@1.8.37 deduped
│ ├── commander@2.9.0 deduped
│ ├── crypto@0.0.3 deduped
│ ├── gdax@0.3.1 deduped
│ ├── limiter@1.1.0 deduped (git://github.com/jhurliman/node-rate-limiter.git#58ce2fda6b5c2bc4ccb81ba3768c5b1bc06c91a5)
│ ├── node.bittrex.api@0.5.1 deduped
│ ├── pushbullet@2.0.0 deduped
│ ├── querystring@0.2.0 deduped
│ ├── simple-mock@0.8.0 deduped
│ ├── superagent@3.0.0 deduped
│ ├── winston@2.3.0 deduped
│ └── ws@1.1.1 deduped
```
Regarding:
> And are you placing trades using the REST interface, or a Trader stream? The latter isn't as well tested (but I am working on it this week)

I'm simply trying to run `tutorials/t005_alertTrader.ts` as is, out of the box, i.e.:
```
GTT.Factories.GDAX.getSubscribedFeeds(options, [product]).then((feed: GDAXFeed) => {
     // --------------------------------------------
     // I GET HERE WITH/WITHOUT ADDING ENVS
     // --------------------------------------------
    
    GTT.Core.createTickerTrigger(feed, product)
        .setAction((ticker: TickerMessage) => {
            const currentPrice = ticker.price;
            GTT.Core.createPriceTrigger(feed, product, currentPrice.minus(spread))
                .setAction((event: TickerMessage) => {
                    pushMessage('Price Trigger', `${base} price has fallen and is now ${event.price} ${quote} on ${product} on GDAX`);
                    // --------------------------------------------
                    // I ONLY GET HERE WITHOUT SUPPLYING ENVS BUT TRADES FAIL 
                    // AS ENVS DON'T APPEAR TO BE PASSED DESPITE WHAT GUIDE SAYS
                    // --------------------------------------------
                    submitTrade('buy', '0.01');
                });

            GTT.Core.createPriceTrigger(feed, product, currentPrice.plus(spread))
                .setAction((event: TickerMessage) => {
                    pushMessage('Price Trigger', `${base} price has risen and is now ${event.price} ${quote} on ${product} on GDAX`);
                    // --------------------------------------------
                    // I ONLY GET HERE WITHOUT SUPPLYING ENVS BUT TRADES FAIL 
                    // AS ENVS DON'T APPEAR TO BE PASSED DESPITE WHAT GUIDE SAYS
                    // --------------------------------------------
                    submitTrade('sell', '0.01');
                });
        });

    GTT.Core.createTickerTrigger(feed, product, false)
        .setAction((ticker: TickerMessage) => {
            console.red(GTT.utils.printTicker(ticker, 3));
        });
});
```  

The non-responses/timeouts occur when I hit  the calls to `GTT.Core.createTickerTrigger(feed, product)`.  When I do not provide api authentication credentials and run the tutorial as is, any attempted trade fails.  However, whenever I add my ENVs to the `GDAXFeedConfig` object, the calls to `GTT.Core.createTickerTrigger(feed, product)` never seem to call back, which is what I assume leads to the timeout/closed connection.

I'm going to try to wipe my npm installation and start from scratch to see if that is the ticket.  Then will attempt to dig deeper into what's occurring with `createTickerTrigger()` for additional debugging and will report back here.

**Comment by kerryknight** [comment]
Well, I have good and bad news.  Good news: running `t005_alertTrader.ts` out of the box is working now when I provide my authentication ENVs.  Bad news: I have no idea *why* it decided to start working now as I did not change a single thing on my end from what had been failing to run previously.  I simply re-ran my existing local copy of the `t005_alertTrader.ts` script again for the upteenth time and it finally began to run with ticker info and attempted trades.  Perhaps it was related to the GDAX instability mentioned by @CjS77? Perhaps it was a wifi/router/local system issue on my end?  No clue. ¯\_(ツ)_/¯

Whatever it was, I'm closing this issue now but will certainly add additional details as available.  Thanks for the assistance!
